cryptographic-message-syntax/src/asn1/rfc5652.rs

```diff
@@ -652,10 +652,18 @@ impl SignedAttributes {
     /// Obtain an instance where the attributes are sorted according to DER
     /// rules. See the comment in [SignerInfo::signed_attributes_digested_content].
     pub fn as_sorted(&self) -> Result<Self, std::io::Error> {
-        // The official rules say you sort by the first element in the container.
-        // That's an Oid, which implements comparisons. So our job is easy.
         let mut res = self.0.clone();
-        res.sort_by(|a, b| a.typ.as_ref().cmp(b.typ.as_ref()));
+        // Ordering on the encoded bytes.
+        // Following: https://stackoverflow.com/a/72492744/2037998
+        res.sort_by_cached_key(|a| {
+            let mut a_encoded = Vec::new();
+            let a_values = a.clone().encode();
+            a_values.write_encoded(Mode::Der, &mut a_encoded).unwrap();
+            a_encoded
+        });
+        // PREVIOUS METHOD: The official rules say you sort by the first element in the container.
```

Comment:
I'm really curious why I left that `first element in the container` comment, which is apparently wrong per your SO discussion. I must have read that somewhere otherwise I wouldn't leave the comment!

Reply:
Yeah I also though this was correct. But apparently not. Also changing this actually fixed some issue with the PDF. So Acrobat can not handle the incorrect sorting.
I did look up if I could find more info somewhere about this. But info was very hard to find and/or vague in its description.
Maybe there are different way of doing this because the standard is not so explicit.

This is what I found around this:
> 11.6 Set-of components
> The encodings of the component values of a set-of value shall appear in ascending order, the encodings being compared as
> octet strings with the shorter components being padded at their trailing end with 0-octets.

Source: https://www.itu.int/rec/T-REC-X.690-202102-I/en

But also:
> 9.3 Set components
> The encodings of the component values of a set value shall appear in an order determined by their tags as specified in 8.6 of
> Rec. ITU-T X.680 | ISO/IEC 8824-1. Additionally, for the purposes of determining the order in which components are
> encoded when one or more component is an untagged choice type, each untagged choice type is ordered as though it has a tag
> equal to that of the smallest tag in that choice type or any untagged choice types nested within.

Source: https://www.itu.int/rec/T-REC-X.690-202102-I/en

Although it looks like `SET` is used for some reason `SET OF` sorting is needed.
```
> openssl asn1parse -in signed_content.der -inform DER
    0:d=0  hl=2 l= 105 cons: SET               
    2:d=1  hl=2 l=  24 cons: SEQUENCE          
    4:d=2  hl=2 l=   9 prim: OBJECT            :contentType
   15:d=2  hl=2 l=  11 cons: SET               
   17:d=3  hl=2 l=   9 prim: OBJECT            :pkcs7-data
   28:d=1  hl=2 l=  28 cons: SEQUENCE          
   30:d=2  hl=2 l=   9 prim: OBJECT            :signingTime
   41:d=2  hl=2 l=  15 cons: SET               
   43:d=3  hl=2 l=  13 prim: UTCTIME           :220726144320Z
   58:d=1  hl=2 l=  47 cons: SEQUENCE          
   60:d=2  hl=2 l=   9 prim: OBJECT            :messageDigest
   71:d=2  hl=2 l=  34 cons: SET               
   73:d=3  hl=2 l=  32 prim: OCTET STRING      [HEX DUMP]
```

This looks to be the clearest source I could find.
> SET OF encoding
> A SET OF items is encoded the same way as a SET, including the tag byte of 0x31. For DER encoding, there is a similar requirement that the SET OF must be encoded in ascending order. Because all elements in the SET OF have the same type, ordering by tag is not sufficient. So the elements of a SET OF are sorted by their encoded values, with shorter values treated as if they were padded to the right with zeroes.

https://letsencrypt.org/docs/a-warm-welcome-to-asn1-and-der/#set-of-encoding

I'm by no means an expert on ASN.1 or DER encoding. But it looks like it works and other people (on StackOverflow) say it should be that way. So I guess that is what is needed.
It could also just be a bug in Acrobat (what would not surprise me, there code seems to hanging by a thread.)

Maybe there is a way to switch between the 2 sorting methods using a variable (that we can implement).
But I don't know what would be best in this case.

Reply:
Thanks for the research! But I'm having a hard time believing I translated any of that into my comment of `sort by the first element in the container`! That being said, I didn't have a great understanding about ASN.1 when I started implementing this crate. So it's quite possible I did something very foolish like interpret `order determined by their tags` to somehow mean `tag = OID`. 🤷‍♂️ 

It doesn't really matter. What does matter is the current sorting is objectively wrong. So we should fix it.